bpo-31567: Use decorator markup for staticmethod()/classmethod()

dmiyakawa opened the pull request:
staticmethod() and classmethod() are expected to be used with decorator, so documentation should mention it.

<!-- issue-number: bpo-31567 -->
https://bugs.python.org/issue31567
<!-- /issue-number -->

Commit:
bpo-31567: Use decorator markup for staticmethod()/classmethod()

merwok commented on Doc/library/functions.rst:
```diff
@@ -196,7 +196,7 @@ are always available.  They are listed here in alphabetical order.
 
 
-.. function:: classmethod(function)
+.. decorator:: classmethod(function)
```
I suspect that this should be `decorator: classmethod`; the version with arguments is probably used for functions that are decorator factories.

(For example, `@functools.lru_cache(max_size=20)` gets arguments and returns a decorator which is then applied to the decorated function; the markup must show clearly if people are expected to call `@decorator` or `@decorator_factory()`.)

dmiyakawa commented:
I have made the requested changes; please review again.

Removed "(function)" for both classmethod and staticmethod. Since the following explanations points to the name "function" in the arg, I modified the description too (I mimicked the line in @functools.singledispatch)

https://docs.python.org/3.6/library/functools.html#functools.singledispatch

> Transforms a function into a single-dispatch generic function.

merwok committed:
Cosmetic edit

Use command form like in docstrings

merwok commented:
Thanks!  Sphinx docs for the reference: http://www.sphinx-doc.org/en/1.6.4/domains.html#directive-py:decorator

I found three misuses of the markup (including singledispatch!) which I will fix in this PR, given that the original ticket was not only about staticmethod and classmethod.

merwok committed:
fix decorator markup

merwok commented:
I wasn’t able to push my commit to your fork (I expected that to be possible given that I could add a commit within github), and I can’t find your fork in the widget to create cross-forks PRs.

https://help.github.com/articles/committing-changes-to-a-pull-request-branch-created-from-a-fork/

dmiyakawa commented:
Sorry I'm not sure what is going on...

I can see you actually added some changes in my PR, I I understand you can add more commits here
https://github.com/dmiyakawa/cpython/commit/9b806c146712c7e05ffcf49138461e73df3e2d76

"Allow edits from maintainers" check  in this page looks already on.
https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/

merwok commented:
Yes that’s the one in created from github.  I made another commit on my laptop and can’t push it to your fork or make a PR: https://github.com/merwok/cpython/commit/59bc4a7025a37d5571adfdd6ed0f1c890d85faca

dmiyakawa committed:
Merge pull request #1 from merwok/fix-issue-31567

fix decorator markup

dmiyakawa commented:
Tried merging the change in merwok/cpython to dmiyakawa/cpython.

miss-islington commented:
Thanks @dmiyakawa for the PR, and @merwok for merging it 🌮🎉.. I'm working now to backport this PR to: 2.7, 3.6.
🐍🍒⛏🤖

miss-islington commented:
Sorry, @dmiyakawa and @merwok, I could not cleanly backport this to `2.7` due to a conflict. 
                            Please backport using [cherry_picker](https://pypi.org/project/cherry-picker/) on command line.
                            ```
                            cherry_picker 0e61e67a57deb4abc677808201d7cf3c38138e02 2.7
                            ```

merwok commented:
I’ll leave 2.7 alone, nothing already uses the decorator markup (maybe it’s not even supported in the sphinx version used by 2.7).

merwok commented:
FTR I think the push issue was due to using an https URL for your fork rather than ssh!
I just pushed commits to the backport PR owned by the miss-islington bot.
